{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# first, import packages\n",
    "import polars as pl\n",
    "import pandas as pd\n",
    "import marimo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Processing of Raw List of Gene Fusion Transcripts from Arriba and FusionCatcher on CCLE+internal Cell Lines**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook details the processes (semi-automated) done to further process the raw output files from Arriba and FusionCatcher fusion transcript callers. \n",
    "\n",
    "1. Run the `pypolars-process-ft-tsv.py` script to generate fusion transcript list from Arriba and FusionCatcher output files. The script takes a mandatory input of path to the directory where sample-specific fusion call output files from Arriba or FusionCatcher are stored as the first argument, and the specific string that is used to identify tool name (`arr` for Arriba fusion transcript call output file prefix, for instance). \n",
    "\n",
    "\tFor example:\n",
    "\t> ``` pypolars-process-ft-tsv.py data/FTmyBRCAs_raw/Arriba arr ```\n",
    "\n",
    "\tDo the same for the FusionCatcher raw output files, as well as the same Arriba and FusionCatcher output files generated from the processing 113 TCGA-Normals (to use as a panel of normals for FT filtering).\n",
    "\n",
    "2. Then, load up the two datasets on Jupyter Notebook and concatenate the dataframes together so that Arriba+FusionCatcher unfiltered FT data are combined into one data table and saved in one `.parquet` and `.tsv` file. Do the same for the `TCGANormals` panel of normals."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load up MyBrCa datasets\n",
    "arr_mdf = pl.scan_parquet('../output/myBRCA/Arriba-FT-all-unfilt-list-v2.parquet')\n",
    "\n",
    "fc_mdf = pl.scan_parquet('../output/myBRCA/FusionCatcher-FT-all-unfilt-list-v2.parquet')\n",
    "\n",
    "# now load TCGANormals\n",
    "arr_tdf = pl.scan_parquet('../output/TCGANormals/Arriba-Normal-FT-all-unfilt-list-v2.parquet')\n",
    "fc_tdf = pl.scan_parquet('../output/TCGANormals/FusionCatcher-Normal-FT-all-unfilt-list-v2.parquet')\n",
    "\n",
    "# html_arr_t = to_html_datatable(pl.DataFrame.to_pandas(arr_tdf.collect(), use_pyarrow_extension_array=True).head(5), display_logo_when_loading=False)\n",
    "\n",
    "# html_fc_t = to_html_datatable(pl.DataFrame.to_pandas(fc_tdf.collect(), use_pyarrow_extension_array=True).head(5), display_logo_when_loading=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loaded Polars dataFrames\n",
    "[//]: # (-.- .tabset .tabset-pills)\n",
    "\n",
    "Here all datasets from the two different fusion transcript calling tools run on both MyBrCa and TCGA-Normals cohorts are shown in tabs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "\n",
       "#### **Dataset 1A** (MyBrCa): Arriba unfiltered\n",
       "Arriba MyBrCa datatable dimension: <b>(49465, 11)</b>\n",
       "\n",
       "Showing truncated table:"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# %%jmd \n",
    "\n",
    "# #### **Dataset 1A** (MyBrCa): Arriba unfiltered\n",
    "# Arriba MyBrCa datatable dimension: <b>{{arr_mdf.collect().shape}}</b>\n",
    "\n",
    "# Showing truncated table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fusionTranscriptID</th>\n",
       "      <th>fusionGeneID</th>\n",
       "      <th>breakpointID</th>\n",
       "      <th>strand1</th>\n",
       "      <th>strand2</th>\n",
       "      <th>site1</th>\n",
       "      <th>site2</th>\n",
       "      <th>type</th>\n",
       "      <th>confidence</th>\n",
       "      <th>sampleID</th>\n",
       "      <th>toolID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TRMT11::SMG6__6:125986622-17:2244719</td>\n",
       "      <td>TRMT11::SMG6</td>\n",
       "      <td>6:125986622-17:2244719</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS/splice-site</td>\n",
       "      <td>CDS/splice-site</td>\n",
       "      <td>translocation</td>\n",
       "      <td>high</td>\n",
       "      <td>1</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STAG3::MEF2C-AS1__7:100189570-5:88919251</td>\n",
       "      <td>STAG3::MEF2C-AS1</td>\n",
       "      <td>7:100189570-5:88919251</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS</td>\n",
       "      <td>intron</td>\n",
       "      <td>translocation/5'-5'</td>\n",
       "      <td>low</td>\n",
       "      <td>1</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MAPK13::C1QL1__6:36132629-17:44965446</td>\n",
       "      <td>MAPK13::C1QL1</td>\n",
       "      <td>6:36132629-17:44965446</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS</td>\n",
       "      <td>intron</td>\n",
       "      <td>translocation/5'-5'</td>\n",
       "      <td>low</td>\n",
       "      <td>1</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STX16::NPEPL1__20:58673711-20:58691724</td>\n",
       "      <td>STX16::NPEPL1</td>\n",
       "      <td>20:58673711-20:58691724</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS/splice-site</td>\n",
       "      <td>5'UTR/splice-site</td>\n",
       "      <td>deletion/read-through</td>\n",
       "      <td>low</td>\n",
       "      <td>1</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MAPK13::NMT1__6:36132629-17:44965446</td>\n",
       "      <td>MAPK13::NMT1</td>\n",
       "      <td>6:36132629-17:44965446</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS</td>\n",
       "      <td>intron</td>\n",
       "      <td>translocation</td>\n",
       "      <td>low</td>\n",
       "      <td>1</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49460</th>\n",
       "      <td>DENND5B::AC087311.1(22711),SYT10(357773)__12:3...</td>\n",
       "      <td>DENND5B::AC087311.1(22711),SYT10(357773)</td>\n",
       "      <td>12:31479608-12:33016465</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>inversion</td>\n",
       "      <td>low</td>\n",
       "      <td>992</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49461</th>\n",
       "      <td>LINC01145::AC245100.2__1:145201150-1:148436753</td>\n",
       "      <td>LINC01145::AC245100.2</td>\n",
       "      <td>1:145201150-1:148436753</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>exon</td>\n",
       "      <td>duplication/5'-5'</td>\n",
       "      <td>low</td>\n",
       "      <td>992</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49462</th>\n",
       "      <td>NET1::RNF169__10:5412820-11:74834676</td>\n",
       "      <td>NET1::RNF169</td>\n",
       "      <td>10:5412820-11:74834676</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS/splice-site</td>\n",
       "      <td>CDS/splice-site</td>\n",
       "      <td>translocation</td>\n",
       "      <td>low</td>\n",
       "      <td>992</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49463</th>\n",
       "      <td>MAN2C1::SIN3A__15:75366522-15:75375872</td>\n",
       "      <td>MAN2C1::SIN3A</td>\n",
       "      <td>15:75366522-15:75375872</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS/splice-site</td>\n",
       "      <td>CDS/splice-site</td>\n",
       "      <td>duplication</td>\n",
       "      <td>low</td>\n",
       "      <td>992</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49464</th>\n",
       "      <td>LINC02224::MRPS30-DT__5:44658462-5:44777328</td>\n",
       "      <td>LINC02224::MRPS30-DT</td>\n",
       "      <td>5:44658462-5:44777328</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>exon/splice-site</td>\n",
       "      <td>exon/splice-site</td>\n",
       "      <td>duplication</td>\n",
       "      <td>low</td>\n",
       "      <td>992</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>49465 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      fusionTranscriptID  \\\n",
       "0                   TRMT11::SMG6__6:125986622-17:2244719   \n",
       "1               STAG3::MEF2C-AS1__7:100189570-5:88919251   \n",
       "2                  MAPK13::C1QL1__6:36132629-17:44965446   \n",
       "3                 STX16::NPEPL1__20:58673711-20:58691724   \n",
       "4                   MAPK13::NMT1__6:36132629-17:44965446   \n",
       "...                                                  ...   \n",
       "49460  DENND5B::AC087311.1(22711),SYT10(357773)__12:3...   \n",
       "49461     LINC01145::AC245100.2__1:145201150-1:148436753   \n",
       "49462               NET1::RNF169__10:5412820-11:74834676   \n",
       "49463             MAN2C1::SIN3A__15:75366522-15:75375872   \n",
       "49464        LINC02224::MRPS30-DT__5:44658462-5:44777328   \n",
       "\n",
       "                                   fusionGeneID             breakpointID  \\\n",
       "0                                  TRMT11::SMG6   6:125986622-17:2244719   \n",
       "1                              STAG3::MEF2C-AS1   7:100189570-5:88919251   \n",
       "2                                 MAPK13::C1QL1   6:36132629-17:44965446   \n",
       "3                                 STX16::NPEPL1  20:58673711-20:58691724   \n",
       "4                                  MAPK13::NMT1   6:36132629-17:44965446   \n",
       "...                                         ...                      ...   \n",
       "49460  DENND5B::AC087311.1(22711),SYT10(357773)  12:31479608-12:33016465   \n",
       "49461                     LINC01145::AC245100.2  1:145201150-1:148436753   \n",
       "49462                              NET1::RNF169   10:5412820-11:74834676   \n",
       "49463                             MAN2C1::SIN3A  15:75366522-15:75375872   \n",
       "49464                      LINC02224::MRPS30-DT    5:44658462-5:44777328   \n",
       "\n",
       "      strand1 strand2             site1              site2  \\\n",
       "0           +       -   CDS/splice-site    CDS/splice-site   \n",
       "1           +       -               CDS             intron   \n",
       "2           +       +               CDS             intron   \n",
       "3           +       +   CDS/splice-site  5'UTR/splice-site   \n",
       "4           +       +               CDS             intron   \n",
       "...       ...     ...               ...                ...   \n",
       "49460       -       +               CDS         intergenic   \n",
       "49461       -       -              exon               exon   \n",
       "49462       +       +   CDS/splice-site    CDS/splice-site   \n",
       "49463       -       -   CDS/splice-site    CDS/splice-site   \n",
       "49464       -       -  exon/splice-site   exon/splice-site   \n",
       "\n",
       "                        type confidence sampleID  toolID  \n",
       "0              translocation       high        1  Arriba  \n",
       "1        translocation/5'-5'        low        1  Arriba  \n",
       "2        translocation/5'-5'        low        1  Arriba  \n",
       "3      deletion/read-through        low        1  Arriba  \n",
       "4              translocation        low        1  Arriba  \n",
       "...                      ...        ...      ...     ...  \n",
       "49460              inversion        low      992  Arriba  \n",
       "49461      duplication/5'-5'        low      992  Arriba  \n",
       "49462          translocation        low      992  Arriba  \n",
       "49463            duplication        low      992  Arriba  \n",
       "49464            duplication        low      992  Arriba  \n",
       "\n",
       "[49465 rows x 11 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# -.-|m { input: false, output: true}\n",
    "pl.DataFrame.to_pandas(arr_mdf.collect(), use_pyarrow_extension_array=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%jmd\n",
    "\n",
    "# #### **Dataset 1B** (MyBrCa): FusionCatcher unfiltered\n",
    "# Arriba MyBrCa datatable dimension: <b>{{fc_mdf.collect().shape}}</b>\n",
    "\n",
    "# Showing truncated table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fusionTranscriptID</th>\n",
       "      <th>fusionGeneID</th>\n",
       "      <th>breakpointID</th>\n",
       "      <th>strand1</th>\n",
       "      <th>strand2</th>\n",
       "      <th>site1</th>\n",
       "      <th>site2</th>\n",
       "      <th>type</th>\n",
       "      <th>confidence</th>\n",
       "      <th>sampleID</th>\n",
       "      <th>toolID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDT2::TAGLN__11:117195915-11:117203002</td>\n",
       "      <td>SIDT2::TAGLN</td>\n",
       "      <td>11:117195915-11:117203002</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS(truncated)</td>\n",
       "      <td>UTR</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>2</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AZGP1::GJC3__7:99971746-7:99923603</td>\n",
       "      <td>AZGP1::GJC3</td>\n",
       "      <td>7:99971746-7:99923603</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>in-frame</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>2</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NPEPPS::TBC1D3__17:47592545-17:38191030</td>\n",
       "      <td>NPEPPS::TBC1D3</td>\n",
       "      <td>17:47592545-17:38191030</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS(complete)</td>\n",
       "      <td>UTR</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>2</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CYP4F11::CYP4F23P__19:15914762-19:15583580</td>\n",
       "      <td>CYP4F11::CYP4F23P</td>\n",
       "      <td>19:15914762-19:15583580</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS(truncated)</td>\n",
       "      <td>exonic(no-known-CDS)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>2</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SLC49A3::ATP5ME__4:686532-4:673401</td>\n",
       "      <td>SLC49A3::ATP5ME</td>\n",
       "      <td>4:686532-4:673401</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>out-of-frame</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>3</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31359</th>\n",
       "      <td>CTBS::GNG5__1:84563257-1:84501970</td>\n",
       "      <td>CTBS::GNG5</td>\n",
       "      <td>1:84563257-1:84501970</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>in-frame</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>991</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31360</th>\n",
       "      <td>MRPS30-DT::LINC02224__5:44808642-5:44658557</td>\n",
       "      <td>MRPS30-DT::LINC02224</td>\n",
       "      <td>5:44808642-5:44658557</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>exonic(no-known-CDS)</td>\n",
       "      <td>exonic(no-known-CDS)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>991</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31361</th>\n",
       "      <td>NBEA::CR382287.1__13:35070852-21:10127330</td>\n",
       "      <td>NBEA::CR382287.1</td>\n",
       "      <td>13:35070852-21:10127330</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS(truncated)</td>\n",
       "      <td>exonic(no-known-CDS)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>991</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31362</th>\n",
       "      <td>HACL1::COLQ__3:15563358-3:15489637</td>\n",
       "      <td>HACL1::COLQ</td>\n",
       "      <td>3:15563358-3:15489637</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>out-of-frame</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>991</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31363</th>\n",
       "      <td>PRCP::TASOR2__11:82900235-10:5739618</td>\n",
       "      <td>PRCP::TASOR2</td>\n",
       "      <td>11:82900235-10:5739618</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>out-of-frame</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>992</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>31364 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                fusionTranscriptID          fusionGeneID  \\\n",
       "0          SIDT2::TAGLN__11:117195915-11:117203002          SIDT2::TAGLN   \n",
       "1               AZGP1::GJC3__7:99971746-7:99923603           AZGP1::GJC3   \n",
       "2          NPEPPS::TBC1D3__17:47592545-17:38191030        NPEPPS::TBC1D3   \n",
       "3       CYP4F11::CYP4F23P__19:15914762-19:15583580     CYP4F11::CYP4F23P   \n",
       "4               SLC49A3::ATP5ME__4:686532-4:673401       SLC49A3::ATP5ME   \n",
       "...                                            ...                   ...   \n",
       "31359            CTBS::GNG5__1:84563257-1:84501970            CTBS::GNG5   \n",
       "31360  MRPS30-DT::LINC02224__5:44808642-5:44658557  MRPS30-DT::LINC02224   \n",
       "31361    NBEA::CR382287.1__13:35070852-21:10127330      NBEA::CR382287.1   \n",
       "31362           HACL1::COLQ__3:15563358-3:15489637           HACL1::COLQ   \n",
       "31363         PRCP::TASOR2__11:82900235-10:5739618          PRCP::TASOR2   \n",
       "\n",
       "                    breakpointID strand1 strand2                 site1  \\\n",
       "0      11:117195915-11:117203002       +       +        CDS(truncated)   \n",
       "1          7:99971746-7:99923603       -       -              in-frame   \n",
       "2        17:47592545-17:38191030       +       -         CDS(complete)   \n",
       "3        19:15914762-19:15583580       -       +        CDS(truncated)   \n",
       "4              4:686532-4:673401       -       -          out-of-frame   \n",
       "...                          ...     ...     ...                   ...   \n",
       "31359      1:84563257-1:84501970       -       -              in-frame   \n",
       "31360      5:44808642-5:44658557       -       -  exonic(no-known-CDS)   \n",
       "31361    13:35070852-21:10127330       +       +        CDS(truncated)   \n",
       "31362      3:15563358-3:15489637       -       -          out-of-frame   \n",
       "31363     11:82900235-10:5739618       -       +          out-of-frame   \n",
       "\n",
       "                      site2 type confidence sampleID         toolID  \n",
       "0                       UTR    .          .        2  FusionCatcher  \n",
       "1                         .    .          .        2  FusionCatcher  \n",
       "2                       UTR    .          .        2  FusionCatcher  \n",
       "3      exonic(no-known-CDS)    .          .        2  FusionCatcher  \n",
       "4                         .    .          .        3  FusionCatcher  \n",
       "...                     ...  ...        ...      ...            ...  \n",
       "31359                     .    .          .      991  FusionCatcher  \n",
       "31360  exonic(no-known-CDS)    .          .      991  FusionCatcher  \n",
       "31361  exonic(no-known-CDS)    .          .      991  FusionCatcher  \n",
       "31362                     .    .          .      991  FusionCatcher  \n",
       "31363                     .    .          .      992  FusionCatcher  \n",
       "\n",
       "[31364 rows x 11 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# -.-|m { input: false, output: true}\n",
    "pl.DataFrame.to_pandas(fc_mdf.collect(), use_pyarrow_extension_array=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "\n",
       "#### **Dataset 2A** (TCGA Normals): FusionCatcher unfiltered\n",
       "Arriba TCGA datatable dimension: <b>(10802, 11)</b>\n",
       "\n",
       "Showing truncated table:"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# %%jmd \n",
    "\n",
    "# #### **Dataset 2A** (TCGA Normals): FusionCatcher unfiltered\n",
    "# Arriba TCGA datatable dimension: <b>{{arr_tdf.collect().shape}}</b>\n",
    "\n",
    "# Showing truncated table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fusionTranscriptID</th>\n",
       "      <th>fusionGeneID</th>\n",
       "      <th>breakpointID</th>\n",
       "      <th>strand1</th>\n",
       "      <th>strand2</th>\n",
       "      <th>site1</th>\n",
       "      <th>site2</th>\n",
       "      <th>type</th>\n",
       "      <th>confidence</th>\n",
       "      <th>sampleID</th>\n",
       "      <th>toolID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TRPM7::SPPL2A__15:50686531-15:50749746</td>\n",
       "      <td>TRPM7::SPPL2A</td>\n",
       "      <td>15:50686531-15:50749746</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS/splice-site</td>\n",
       "      <td>CDS/splice-site</td>\n",
       "      <td>duplication</td>\n",
       "      <td>high</td>\n",
       "      <td>TCGA-A7-A0CE</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AC084756.2::SPPL2A__15:50686531-15:50749746</td>\n",
       "      <td>AC084756.2::SPPL2A</td>\n",
       "      <td>15:50686531-15:50749746</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS/splice-site</td>\n",
       "      <td>CDS/splice-site</td>\n",
       "      <td>duplication</td>\n",
       "      <td>high</td>\n",
       "      <td>TCGA-A7-A0CE</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BOLA2B::SMG1P5__16:30193358-16:30288681</td>\n",
       "      <td>BOLA2B::SMG1P5</td>\n",
       "      <td>16:30193358-16:30288681</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS/splice-site</td>\n",
       "      <td>exon</td>\n",
       "      <td>duplication</td>\n",
       "      <td>high</td>\n",
       "      <td>TCGA-A7-A0CE</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FBXO25::SEPTIN14__8:435707-7:55796092</td>\n",
       "      <td>FBXO25::SEPTIN14</td>\n",
       "      <td>8:435707-7:55796092</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS/splice-site</td>\n",
       "      <td>CDS/splice-site</td>\n",
       "      <td>translocation</td>\n",
       "      <td>high</td>\n",
       "      <td>TCGA-A7-A0CE</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TVP23C::CDRT4__17:15503098-17:15440285</td>\n",
       "      <td>TVP23C::CDRT4</td>\n",
       "      <td>17:15503098-17:15440285</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS</td>\n",
       "      <td>5'UTR/splice-site</td>\n",
       "      <td>deletion/read-through</td>\n",
       "      <td>medium</td>\n",
       "      <td>TCGA-A7-A0CE</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10797</th>\n",
       "      <td>PRSS45P::PRSS50(438),PRSS46P(1276)__3:46742488...</td>\n",
       "      <td>PRSS45P::PRSS50(438),PRSS46P(1276)</td>\n",
       "      <td>3:46742488-3:46718307</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>deletion/read-through</td>\n",
       "      <td>low</td>\n",
       "      <td>TCGA-GI-A2C9</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10798</th>\n",
       "      <td>LRRK1::AC090907.2(4523),AC019254.2(26014)__15:...</td>\n",
       "      <td>LRRK1::AC090907.2(4523),AC019254.2(26014)</td>\n",
       "      <td>15:101049667-15:101090589</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>deletion/read-through</td>\n",
       "      <td>low</td>\n",
       "      <td>TCGA-GI-A2C9</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10799</th>\n",
       "      <td>PTPRT::AL049812.2__20:42081907-20:42052835</td>\n",
       "      <td>PTPRT::AL049812.2</td>\n",
       "      <td>20:42081907-20:42052835</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS</td>\n",
       "      <td>intron</td>\n",
       "      <td>deletion/read-through/5'-5'</td>\n",
       "      <td>low</td>\n",
       "      <td>TCGA-GI-A2C9</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10800</th>\n",
       "      <td>LINC01128::PDCD2(26680),AL031259.1(1250)__1:82...</td>\n",
       "      <td>LINC01128::PDCD2(26680),AL031259.1(1250)</td>\n",
       "      <td>1:828888-6:170611372</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>intron</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>translocation</td>\n",
       "      <td>low</td>\n",
       "      <td>TCGA-GI-A2C9</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10801</th>\n",
       "      <td>SNX12::RNU1-56P(19690),SOCS5P4(3086)__X:710610...</td>\n",
       "      <td>SNX12::RNU1-56P(19690),SOCS5P4(3086)</td>\n",
       "      <td>X:71061025-X:71040128</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>deletion/read-through</td>\n",
       "      <td>low</td>\n",
       "      <td>TCGA-GI-A2C9</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10802 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      fusionTranscriptID  \\\n",
       "0                 TRPM7::SPPL2A__15:50686531-15:50749746   \n",
       "1            AC084756.2::SPPL2A__15:50686531-15:50749746   \n",
       "2                BOLA2B::SMG1P5__16:30193358-16:30288681   \n",
       "3                  FBXO25::SEPTIN14__8:435707-7:55796092   \n",
       "4                 TVP23C::CDRT4__17:15503098-17:15440285   \n",
       "...                                                  ...   \n",
       "10797  PRSS45P::PRSS50(438),PRSS46P(1276)__3:46742488...   \n",
       "10798  LRRK1::AC090907.2(4523),AC019254.2(26014)__15:...   \n",
       "10799         PTPRT::AL049812.2__20:42081907-20:42052835   \n",
       "10800  LINC01128::PDCD2(26680),AL031259.1(1250)__1:82...   \n",
       "10801  SNX12::RNU1-56P(19690),SOCS5P4(3086)__X:710610...   \n",
       "\n",
       "                                    fusionGeneID               breakpointID  \\\n",
       "0                                  TRPM7::SPPL2A    15:50686531-15:50749746   \n",
       "1                             AC084756.2::SPPL2A    15:50686531-15:50749746   \n",
       "2                                 BOLA2B::SMG1P5    16:30193358-16:30288681   \n",
       "3                               FBXO25::SEPTIN14        8:435707-7:55796092   \n",
       "4                                  TVP23C::CDRT4    17:15503098-17:15440285   \n",
       "...                                          ...                        ...   \n",
       "10797         PRSS45P::PRSS50(438),PRSS46P(1276)      3:46742488-3:46718307   \n",
       "10798  LRRK1::AC090907.2(4523),AC019254.2(26014)  15:101049667-15:101090589   \n",
       "10799                          PTPRT::AL049812.2    20:42081907-20:42052835   \n",
       "10800   LINC01128::PDCD2(26680),AL031259.1(1250)       1:828888-6:170611372   \n",
       "10801       SNX12::RNU1-56P(19690),SOCS5P4(3086)      X:71061025-X:71040128   \n",
       "\n",
       "      strand1 strand2            site1              site2  \\\n",
       "0           -       -  CDS/splice-site    CDS/splice-site   \n",
       "1           -       -  CDS/splice-site    CDS/splice-site   \n",
       "2           -       -  CDS/splice-site               exon   \n",
       "3           +       -  CDS/splice-site    CDS/splice-site   \n",
       "4           -       -              CDS  5'UTR/splice-site   \n",
       "...       ...     ...              ...                ...   \n",
       "10797       -       -              CDS         intergenic   \n",
       "10798       +       +              CDS         intergenic   \n",
       "10799       -       -              CDS             intron   \n",
       "10800       .       .           intron         intergenic   \n",
       "10801       -       -              CDS         intergenic   \n",
       "\n",
       "                              type confidence      sampleID  toolID  \n",
       "0                      duplication       high  TCGA-A7-A0CE  Arriba  \n",
       "1                      duplication       high  TCGA-A7-A0CE  Arriba  \n",
       "2                      duplication       high  TCGA-A7-A0CE  Arriba  \n",
       "3                    translocation       high  TCGA-A7-A0CE  Arriba  \n",
       "4            deletion/read-through     medium  TCGA-A7-A0CE  Arriba  \n",
       "...                            ...        ...           ...     ...  \n",
       "10797        deletion/read-through        low  TCGA-GI-A2C9  Arriba  \n",
       "10798        deletion/read-through        low  TCGA-GI-A2C9  Arriba  \n",
       "10799  deletion/read-through/5'-5'        low  TCGA-GI-A2C9  Arriba  \n",
       "10800                translocation        low  TCGA-GI-A2C9  Arriba  \n",
       "10801        deletion/read-through        low  TCGA-GI-A2C9  Arriba  \n",
       "\n",
       "[10802 rows x 11 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# -.-|m { input: false, output: true}\n",
    "pl.DataFrame.to_pandas(arr_tdf.collect(), use_pyarrow_extension_array=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "\n",
       "#### **Dataset 2B** (TCGA Normals): FusionCatcher unfiltered\n",
       "FusionCatcher TCGA Normals datatable dimension: <b>(3759, 11)</b>\n",
       "\n",
       "Showing truncated table:"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# %%jmd\n",
    "\n",
    "# #### **Dataset 2B** (TCGA Normals): FusionCatcher unfiltered\n",
    "# FusionCatcher TCGA Normals datatable dimension: <b>{{fc_tdf.collect().shape}}</b>\n",
    "\n",
    "# Showing truncated table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fusionTranscriptID</th>\n",
       "      <th>fusionGeneID</th>\n",
       "      <th>breakpointID</th>\n",
       "      <th>strand1</th>\n",
       "      <th>strand2</th>\n",
       "      <th>site1</th>\n",
       "      <th>site2</th>\n",
       "      <th>type</th>\n",
       "      <th>confidence</th>\n",
       "      <th>sampleID</th>\n",
       "      <th>toolID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AZGP1::GJC3__7:99971746-7:99923603</td>\n",
       "      <td>AZGP1::GJC3</td>\n",
       "      <td>7:99971746-7:99923603</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>in-frame</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>TCGA-A7-A0CE</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AZGP1::GJC3__7:99971745-7:99929620</td>\n",
       "      <td>AZGP1::GJC3</td>\n",
       "      <td>7:99971745-7:99929620</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>UTR</td>\n",
       "      <td>CDS(complete)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>TCGA-A7-A0CE</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NPEPPS::TBC1D3__17:47592545-17:38191030</td>\n",
       "      <td>NPEPPS::TBC1D3</td>\n",
       "      <td>17:47592545-17:38191030</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS(complete)</td>\n",
       "      <td>UTR</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>TCGA-A7-A0CE</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SMG1::NPIPB5__16:18858170-16:22513522</td>\n",
       "      <td>SMG1::NPIPB5</td>\n",
       "      <td>16:18858170-16:22513522</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS(truncated)</td>\n",
       "      <td>UTR</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>TCGA-A7-A0CE</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SMG1::NPIPB5__16:18858211-16:22513522</td>\n",
       "      <td>SMG1::NPIPB5</td>\n",
       "      <td>16:18858211-16:22513522</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS(truncated)</td>\n",
       "      <td>UTR</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>TCGA-A7-A0CE</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3754</th>\n",
       "      <td>F11R::RBM4__1:161019399-11:66652310</td>\n",
       "      <td>F11R::RBM4</td>\n",
       "      <td>1:161019399-11:66652310</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>intronic</td>\n",
       "      <td>CDS(truncated)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>TCGA-GI-A2C9</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3755</th>\n",
       "      <td>FBXO25::FAM157B__8:435707-9:138243620</td>\n",
       "      <td>FBXO25::FAM157B</td>\n",
       "      <td>8:435707-9:138243620</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS(truncated)</td>\n",
       "      <td>exonic(no-known-CDS)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>TCGA-GI-A2C9</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3756</th>\n",
       "      <td>MAGT1::STAC2__X:77891455-17:39211643</td>\n",
       "      <td>MAGT1::STAC2</td>\n",
       "      <td>X:77891455-17:39211643</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>intronic</td>\n",
       "      <td>UTR</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>TCGA-GI-A2C9</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3757</th>\n",
       "      <td>PTPRF::IGK@__1:43622869-2:89150246</td>\n",
       "      <td>PTPRF::IGK@</td>\n",
       "      <td>1:43622869-2:89150246</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>UTR</td>\n",
       "      <td>---</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>TCGA-GI-A2C9</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3758</th>\n",
       "      <td>EEF1AKMT3::TSFM__12:57773128-12:57786163</td>\n",
       "      <td>EEF1AKMT3::TSFM</td>\n",
       "      <td>12:57773128-12:57786163</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>in-frame</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>TCGA-GI-A2C9</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3759 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            fusionTranscriptID     fusionGeneID  \\\n",
       "0           AZGP1::GJC3__7:99971746-7:99923603      AZGP1::GJC3   \n",
       "1           AZGP1::GJC3__7:99971745-7:99929620      AZGP1::GJC3   \n",
       "2      NPEPPS::TBC1D3__17:47592545-17:38191030   NPEPPS::TBC1D3   \n",
       "3        SMG1::NPIPB5__16:18858170-16:22513522     SMG1::NPIPB5   \n",
       "4        SMG1::NPIPB5__16:18858211-16:22513522     SMG1::NPIPB5   \n",
       "...                                        ...              ...   \n",
       "3754       F11R::RBM4__1:161019399-11:66652310       F11R::RBM4   \n",
       "3755     FBXO25::FAM157B__8:435707-9:138243620  FBXO25::FAM157B   \n",
       "3756      MAGT1::STAC2__X:77891455-17:39211643     MAGT1::STAC2   \n",
       "3757        PTPRF::IGK@__1:43622869-2:89150246      PTPRF::IGK@   \n",
       "3758  EEF1AKMT3::TSFM__12:57773128-12:57786163  EEF1AKMT3::TSFM   \n",
       "\n",
       "                 breakpointID strand1 strand2           site1  \\\n",
       "0       7:99971746-7:99923603       -       -        in-frame   \n",
       "1       7:99971745-7:99929620       -       -             UTR   \n",
       "2     17:47592545-17:38191030       +       -   CDS(complete)   \n",
       "3     16:18858170-16:22513522       -       +  CDS(truncated)   \n",
       "4     16:18858211-16:22513522       -       +  CDS(truncated)   \n",
       "...                       ...     ...     ...             ...   \n",
       "3754  1:161019399-11:66652310       -       +        intronic   \n",
       "3755     8:435707-9:138243620       +       +  CDS(truncated)   \n",
       "3756   X:77891455-17:39211643       -       -        intronic   \n",
       "3757    1:43622869-2:89150246       +       -             UTR   \n",
       "3758  12:57773128-12:57786163       +       +        in-frame   \n",
       "\n",
       "                     site2 type confidence      sampleID         toolID  \n",
       "0                        .    .          .  TCGA-A7-A0CE  FusionCatcher  \n",
       "1            CDS(complete)    .          .  TCGA-A7-A0CE  FusionCatcher  \n",
       "2                      UTR    .          .  TCGA-A7-A0CE  FusionCatcher  \n",
       "3                      UTR    .          .  TCGA-A7-A0CE  FusionCatcher  \n",
       "4                      UTR    .          .  TCGA-A7-A0CE  FusionCatcher  \n",
       "...                    ...  ...        ...           ...            ...  \n",
       "3754        CDS(truncated)    .          .  TCGA-GI-A2C9  FusionCatcher  \n",
       "3755  exonic(no-known-CDS)    .          .  TCGA-GI-A2C9  FusionCatcher  \n",
       "3756                   UTR    .          .  TCGA-GI-A2C9  FusionCatcher  \n",
       "3757                   ---    .          .  TCGA-GI-A2C9  FusionCatcher  \n",
       "3758                     .    .          .  TCGA-GI-A2C9  FusionCatcher  \n",
       "\n",
       "[3759 rows x 11 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# -.-|m { input: false, output: true}\n",
    "pl.DataFrame.to_pandas(fc_tdf.collect(), use_pyarrow_extension_array=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Concatenate Arriba and FusionCatcher Datasets**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we can merge the two dataframes into one masterFrame for each cohort data (MyBrCa & TCGA panel of normals) using Polars' `concat`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**NOTE:** Vertical concatenation is the default, where two dataframes sharing the exact same columns would be joined together, adding all rows of dataframe 1 and 2 vertically.\n",
    "\n",
    "[//]: # (-.- .alert .alert-warning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -.-|m { input: false, output: false}\n",
    "from polars.exceptions import CategoricalRemappingWarning\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore', category=CategoricalRemappingWarning)\n",
    "\n",
    "joined_df = pl.concat(\n",
    "    [\n",
    "        arr_mdf.collect(),\n",
    "        fc_mdf.collect()\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "Concatenated MyBrCa Arriba+FusionCatcher datatable dimension: <b>(80829, 11)</b>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# -.-|m { input: false, output: true}\n",
    "display(HTML(f\"Concatenated MyBrCa Arriba+FusionCatcher datatable dimension: \" + f\"<b>{joined_df.shape}</b>\"))\n",
    "\n",
    "# show(joined_df.head(5), maxBytes=0, classes=\"display compact\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Do the same with the TCGA panel of normal FTs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -.-|m { input: false, output: false}\n",
    "from polars.exceptions import CategoricalRemappingWarning\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore', category=CategoricalRemappingWarning)\n",
    "\n",
    "joined_norms_df = pl.concat(\n",
    "    [\n",
    "        arr_tdf.collect(),\n",
    "        fc_tdf.collect()\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "Concatenated TCGA-Normals Arriba+FusionCatcher datatable dimension: <b>(14561, 11)</b>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# -.-|m { input: false, output: true}\n",
    "\n",
    "display(HTML(f\"Concatenated TCGA-Normals Arriba+FusionCatcher datatable dimension: \" + f\"<b>{joined_norms_df.shape}</b>\"))\n",
    "\n",
    "# show(joined_norms_df.head(5), maxBytes=0, classes=\"display compact\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Filter MyBrCa Merged DataFrame using Panel of Normals**\n",
    "Now we can filter the unfiltered, concatenated FT dataframes by discarding those that are present in TCGA Normal data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -.-|m { input: true, output: true}\n",
    "\n",
    "mybrca_ccdf = pl.scan_parquet('../output/myBRCA/Arr_FC-concat-FT-all-unfilt-list-v2.parquet')\n",
    "\n",
    "tcganorms_ccdf = pl.scan_parquet('../output/TCGANormals/Arr_FC-Normals-concat-FT-all-unfilt-list-v2.parquet')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Once they are loaded, we can convert to Pandas from Polars for ease of processing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -.-|m { input: true, output: true, input_fold: show}\n",
    "\n",
    "my_concat_df = pl.DataFrame.to_pandas(mybrca_ccdf.collect(), use_pyarrow_extension_array=True)\n",
    "tn_concat_df = pl.DataFrame.to_pandas(tcganorms_ccdf.collect(), use_pyarrow_extension_array=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "Concatenated MyBrCa Arriba+FusionCatcher datatable dimension: <b>(80829, 11)</b>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# # -.-|m { input: false, output: true}\n",
    "# display(HTML(f\"Concatenated MyBrCa Arriba+FusionCatcher datatable dimension: \" + f\"<b>{my_concat_df.shape}</b>\"))\n",
    "\n",
    "# show(my_concat_df.head(5), maxBytes=0, classes=\"display compact\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "Concatenated TCGA Normals Arriba+FusionCatcher datatable dimension: <b>(14561, 11)</b>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# -.-|m { input: false, output: true}\n",
    "# display(HTML(f\"Concatenated TCGA Normals Arriba+FusionCatcher datatable dimension: \" + f\"<b>{tn_concat_df.shape}</b>\"))\n",
    "\n",
    "# show(tn_concat_df.head(5), maxBytes=0, classes=\"display compact\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load up test tsvs\n",
    "arr_test_df = pd.read_csv('../data/minimal-test/373T_arr-v113-refannot.tsv', sep='\\t')\n",
    "# arr_test_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fc_test_df = pd.read_csv('../data/minimal-test/373T_fc-v113-refannot.tsv', sep='\\t')\n",
    "# fc_test_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ASIDE: Modify `collate-fts-nf.py` to handle NaN in GeneIDs (in FusionCatcher files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample name: 373T\n",
      "Input arguments: [('/home/ec2-user/repos/FT-NeonDisco/data/minimal-test/373T_arr.tsv', 'arr'), ('/home/ec2-user/repos/FT-NeonDisco/data/minimal-test/373T_fc.tsv', 'fc')]\n",
      "Setting tool name...\n",
      "Reading Arriba of 373T TSV file...(sample ID: 373)\n",
      "Setting tool name...\n",
      "Reading FusionCatcher of 373T TSV file...(sample ID: 373)\n",
      "Concatenating lazy DataFrames from Arriba and FusionCatcher...\n",
      "Concatenation completed. Collecting...\n",
      "Saving as parquet and tsv files...\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "python /home/ec2-user/repos/FT-NeonDisco/scripts/collate-fts-nf-v2.py 373T /home/ec2-user/repos/FT-NeonDisco/data/minimal-test/373T_arr.tsv arr /home/ec2-user/repos/FT-NeonDisco/data/minimal-test/373T_fc.tsv fc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fusionTranscriptID</th>\n",
       "      <th>fusionGenePair</th>\n",
       "      <th>breakpointID</th>\n",
       "      <th>strand1</th>\n",
       "      <th>strand2</th>\n",
       "      <th>site1</th>\n",
       "      <th>site2</th>\n",
       "      <th>type</th>\n",
       "      <th>confidence</th>\n",
       "      <th>sampleID</th>\n",
       "      <th>toolID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SPAG5-AS1::RAB11FIP4__17:28616389-17:31529712</td>\n",
       "      <td>SPAG5-AS1::RAB11FIP4</td>\n",
       "      <td>17:28616389-17:31529712</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>intron</td>\n",
       "      <td>duplication</td>\n",
       "      <td>high</td>\n",
       "      <td>373</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NSF::ENSG00000289599__17:46643259-17:48009736</td>\n",
       "      <td>NSF::ENSG00000289599</td>\n",
       "      <td>17:46643259-17:48009736</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS/splice-site</td>\n",
       "      <td>exon/splice-site</td>\n",
       "      <td>deletion</td>\n",
       "      <td>high</td>\n",
       "      <td>373</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NSF::ENSG00000289599__17:46643259-17:48012093</td>\n",
       "      <td>NSF::ENSG00000289599</td>\n",
       "      <td>17:46643259-17:48012093</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS/splice-site</td>\n",
       "      <td>exon/splice-site</td>\n",
       "      <td>deletion</td>\n",
       "      <td>high</td>\n",
       "      <td>373</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NSF::ENSG00000289599__17:46643259-17:48012140</td>\n",
       "      <td>NSF::ENSG00000289599</td>\n",
       "      <td>17:46643259-17:48012140</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS/splice-site</td>\n",
       "      <td>exon/splice-site</td>\n",
       "      <td>deletion</td>\n",
       "      <td>medium</td>\n",
       "      <td>373</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NSF::ENSG00000289599__17:46643259-17:48011155</td>\n",
       "      <td>NSF::ENSG00000289599</td>\n",
       "      <td>17:46643259-17:48011155</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS/splice-site</td>\n",
       "      <td>exon/splice-site</td>\n",
       "      <td>deletion</td>\n",
       "      <td>medium</td>\n",
       "      <td>373</td>\n",
       "      <td>Arriba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>ENSG00000294461::PITPNB__22:27930694-22:27914347</td>\n",
       "      <td>ENSG00000294461::PITPNB</td>\n",
       "      <td>22:27930694-22:27914347</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>exonic(no-known-CDS)</td>\n",
       "      <td>CDS(truncated)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>373</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>BMS1P14::ENSG00000286129__9:40571871-22:22318899</td>\n",
       "      <td>BMS1P14::ENSG00000286129</td>\n",
       "      <td>9:40571871-22:22318899</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>exonic(no-known-CDS)</td>\n",
       "      <td>exonic(no-known-CDS)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>373</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>GATAD2A::LINC02684__19:19386138-11:135061947</td>\n",
       "      <td>GATAD2A::LINC02684</td>\n",
       "      <td>19:19386138-11:135061947</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>UTR</td>\n",
       "      <td>exonic(no-known-CDS)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>373</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>NAIP::OCLN__5:70979869-5:69534694</td>\n",
       "      <td>NAIP::OCLN</td>\n",
       "      <td>5:70979869-5:69534694</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>out-of-frame</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>373</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>XPO1::ERBB2__2:61514062-17:39715286</td>\n",
       "      <td>XPO1::ERBB2</td>\n",
       "      <td>2:61514062-17:39715286</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>UTR</td>\n",
       "      <td>CDS(truncated)</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>373</td>\n",
       "      <td>FusionCatcher</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>126 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   fusionTranscriptID  \\\n",
       "0       SPAG5-AS1::RAB11FIP4__17:28616389-17:31529712   \n",
       "1       NSF::ENSG00000289599__17:46643259-17:48009736   \n",
       "2       NSF::ENSG00000289599__17:46643259-17:48012093   \n",
       "3       NSF::ENSG00000289599__17:46643259-17:48012140   \n",
       "4       NSF::ENSG00000289599__17:46643259-17:48011155   \n",
       "..                                                ...   \n",
       "121  ENSG00000294461::PITPNB__22:27930694-22:27914347   \n",
       "122  BMS1P14::ENSG00000286129__9:40571871-22:22318899   \n",
       "123      GATAD2A::LINC02684__19:19386138-11:135061947   \n",
       "124                 NAIP::OCLN__5:70979869-5:69534694   \n",
       "125               XPO1::ERBB2__2:61514062-17:39715286   \n",
       "\n",
       "               fusionGenePair              breakpointID strand1 strand2  \\\n",
       "0        SPAG5-AS1::RAB11FIP4   17:28616389-17:31529712       -       -   \n",
       "1        NSF::ENSG00000289599   17:46643259-17:48009736       +       +   \n",
       "2        NSF::ENSG00000289599   17:46643259-17:48012093       +       +   \n",
       "3        NSF::ENSG00000289599   17:46643259-17:48012140       +       +   \n",
       "4        NSF::ENSG00000289599   17:46643259-17:48011155       +       +   \n",
       "..                        ...                       ...     ...     ...   \n",
       "121   ENSG00000294461::PITPNB   22:27930694-22:27914347       -       -   \n",
       "122  BMS1P14::ENSG00000286129    9:40571871-22:22318899       -       +   \n",
       "123        GATAD2A::LINC02684  19:19386138-11:135061947       +       -   \n",
       "124                NAIP::OCLN     5:70979869-5:69534694       -       +   \n",
       "125               XPO1::ERBB2    2:61514062-17:39715286       -       +   \n",
       "\n",
       "                    site1                 site2         type confidence  \\\n",
       "0                    exon                intron  duplication       high   \n",
       "1         CDS/splice-site      exon/splice-site     deletion       high   \n",
       "2         CDS/splice-site      exon/splice-site     deletion       high   \n",
       "3         CDS/splice-site      exon/splice-site     deletion     medium   \n",
       "4         CDS/splice-site      exon/splice-site     deletion     medium   \n",
       "..                    ...                   ...          ...        ...   \n",
       "121  exonic(no-known-CDS)        CDS(truncated)            .          .   \n",
       "122  exonic(no-known-CDS)  exonic(no-known-CDS)            .          .   \n",
       "123                   UTR  exonic(no-known-CDS)            .          .   \n",
       "124          out-of-frame                     .            .          .   \n",
       "125                   UTR        CDS(truncated)            .          .   \n",
       "\n",
       "     sampleID         toolID  \n",
       "0         373         Arriba  \n",
       "1         373         Arriba  \n",
       "2         373         Arriba  \n",
       "3         373         Arriba  \n",
       "4         373         Arriba  \n",
       "..        ...            ...  \n",
       "121       373  FusionCatcher  \n",
       "122       373  FusionCatcher  \n",
       "123       373  FusionCatcher  \n",
       "124       373  FusionCatcher  \n",
       "125       373  FusionCatcher  \n",
       "\n",
       "[126 rows x 11 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load up collated data\n",
    "collated_df = pd.read_csv('../data/minimal-test/373T-collated-FT-UNFILTERED-collate-v2.tsv', sep='\\t')\n",
    "collated_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load up combined parquet then convert to polars\n",
    "combined_pq_df = pl.scan_parquet('../data/minimal-test/373T-collated-FT-UNFILTERED.parquet').collect().to_pandas(use_pyarrow_extension_array=True)\n",
    "combined_pq_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #show only unique breakpointIDs\n",
    "# collated_unique_df = collated_df.drop_duplicates(subset=['breakpointID'])\n",
    "# collated_unique_df.to_csv('../data/minimal-test/373T-collated-FT-UNFILTERED-collate-v2-unique.tsv', sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_pq_df = pd.read_csv('../data/minimal-test/373T-collated-FT-UNFILTERED-collate-v2.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### 1. Get unique breakpointIDs from the combined ft dataframe based on toolID and add detection information\n",
    "\n",
    "# Split the dataframe by toolID\n",
    "arriba_df = combined_pq_df[combined_pq_df['toolID'] == 'Arriba'].copy()\n",
    "fusioncatcher_df = combined_pq_df[combined_pq_df['toolID'] == 'FusionCatcher'].copy()\n",
    "\n",
    "# Drop duplicates by breakpointID for each tool separately\n",
    "arriba_unique_df = arriba_df.drop_duplicates(subset=['breakpointID']).copy()\n",
    "fusioncatcher_unique_df = fusioncatcher_df.drop_duplicates(subset=['breakpointID']).copy()\n",
    "\n",
    "# Get sets of breakpointIDs from each tool\n",
    "arriba_breakpoints = set(arriba_unique_df['breakpointID'])\n",
    "fusioncatcher_breakpoints = set(fusioncatcher_unique_df['breakpointID'])\n",
    "\n",
    "# Find breakpoints detected by both tools\n",
    "common_breakpoints = arriba_breakpoints.intersection(fusioncatcher_breakpoints)\n",
    "\n",
    "# Add detection information using assign() with the new column name\n",
    "arriba_unique_df = arriba_unique_df.assign(detectedBy='Arriba')\n",
    "fusioncatcher_unique_df = fusioncatcher_unique_df.assign(detectedBy='FusionCatcher')\n",
    "\n",
    "# Mark breakpoints found by both tools\n",
    "mask = arriba_unique_df['breakpointID'].isin(common_breakpoints)\n",
    "arriba_unique_df.loc[mask, 'detectedBy'] = 'Both'\n",
    "\n",
    "# Only keep FusionCatcher rows that weren't detected by Arriba\n",
    "fusioncatcher_exclusive_df = fusioncatcher_unique_df[~fusioncatcher_unique_df['breakpointID'].isin(arriba_breakpoints)]\n",
    "\n",
    "# Combine Arriba rows with FusionCatcher-exclusive rows\n",
    "combined_unique_df = pd.concat([arriba_unique_df, fusioncatcher_exclusive_df])\n",
    "\n",
    "# 2. Filter out breakpoints seen in TCGA normals\n",
    "# load up the TCGA normals (all Arr and FC) dataframe\n",
    "panel_normals_uniq_df = pd.read_csv('../output/TCGANormals/Arr-and-FC_TCGANormals-UNIQUE-breakpointID-list.tsv', sep='\\t')\n",
    "\n",
    "# Get the unique breakpointIDs from the panel normals\n",
    "panel_normals_uniq_fts = set(panel_normals_uniq_df['breakpointID'])\n",
    "panel_normals_uniq_fts\n",
    "\n",
    "# Filter out breakpoints seen in TCGA normals\n",
    "final_unique_df = combined_unique_df[~combined_unique_df['breakpointID'].isin(panel_normals_uniq_fts)]\n",
    "\n",
    "# Save the result\n",
    "final_unique_df.to_csv('../output/minimal-test/373T-combined-FT-uniq-FILTERED.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ASIDE: Postprocess TCGA Normal to get unique breakpoints and add detection information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Get unique breakpointIDs from the combined ft dataframe based on toolID and add detection information\n",
    "\n",
    "# load up the TCGA normals (all Arr and FC) dataframe\n",
    "panel_normals_df = pl.scan_parquet('../output/TCGANormals/Arr_FC-Normals-concat-FT-all-unfilt-list-v2.parquet').collect().to_pandas(use_pyarrow_extension_array=True)\n",
    "\n",
    "# Split the dataframe by toolID\n",
    "arriba_norm_df = panel_normals_df[panel_normals_df['toolID'] == 'Arriba'].copy()\n",
    "fusioncatcher_norm_df = panel_normals_df[panel_normals_df['toolID'] == 'FusionCatcher'].copy()\n",
    "\n",
    "# Drop duplicates by breakpointID for each tool separately\n",
    "arriba_norm_unique_df = arriba_norm_df.drop_duplicates(subset=['breakpointID']).copy()\n",
    "fusioncatcher_norm_unique_df = fusioncatcher_norm_df.drop_duplicates(subset=['breakpointID']).copy()\n",
    "\n",
    "# Get sets of breakpointIDs from each tool\n",
    "arriba_norm_breakpoints = set(arriba_norm_unique_df['breakpointID'])\n",
    "fusioncatcher_norm_breakpoints = set(fusioncatcher_norm_unique_df['breakpointID'])\n",
    "\n",
    "# Find breakpoints detected by both tools\n",
    "common_norm_breakpoints = arriba_norm_breakpoints.intersection(fusioncatcher_norm_breakpoints)\n",
    "\n",
    "# Add detection information using assign() with the new column name\n",
    "arriba_norm_unique_df = arriba_norm_unique_df.assign(detectedBy='Arriba')\n",
    "fusioncatcher_norm_unique_df = fusioncatcher_norm_unique_df.assign(detectedBy='FusionCatcher')\n",
    "\n",
    "# Mark breakpoints found by both tools\n",
    "mask = arriba_norm_unique_df['breakpointID'].isin(common_norm_breakpoints)\n",
    "arriba_norm_unique_df.loc[mask, 'detectedBy'] = 'Both'\n",
    "\n",
    "# Only keep FusionCatcher rows that weren't detected by Arriba\n",
    "fusioncatcher_norm_exclusive_df = fusioncatcher_norm_unique_df[~fusioncatcher_norm_unique_df['breakpointID'].isin(arriba_norm_breakpoints)]\n",
    "\n",
    "# Combine Arriba rows with FusionCatcher-exclusive rows\n",
    "combined_norm_unique_df = pd.concat([arriba_norm_unique_df, fusioncatcher_norm_exclusive_df])\n",
    "\n",
    "combined_norm_unique_df\n",
    "\n",
    "\n",
    "# Save the result\n",
    "combined_norm_unique_df.to_csv('../output/TCGANormals/Arr-and-FC_TCGANormals-UNIQUE-breakpointID-list.tsv', sep='\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "datasci",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
